### ISSUE TITLE
--bisect fails for large number of specs

### ISSUE THREAD

**Issue by wendorf** [rspec/rspec-core#2145]
When running `--bisect` with a large number of examples, the process fails with a Errno::E2BIG when attempting to popen_run:

```
Bisect started using options: ""
Running suite to find failures... (8 minutes 53 seconds)
Starting bisect with 2 failing examples and 9851 non-failing examples.
Checking that failure(s) are order-dependent... failure appears to be order-dependent

Round 1: bisecting over non-failing examples 1-9851
/Users/me/.rubies/ruby-2.2.4/lib/ruby/2.2.0/open3.rb:193:in `spawn': Argument list too long - /Users/me/.rubies/ruby-2.2.4/bin/ruby -I[trimmed]  /Users/me/.gem/ruby/2.2.4/gems/rspec-core-3.4.1/exe/rspec --format bisect --drb-port 58073 ./spec/unit/example_spec.rb\[1:8:1:1\] [trimmed many specs] (Errno::E2BIG)
	from /Users/me/.rubies/ruby-2.2.4/lib/ruby/2.2.0/open3.rb:193:in `popen_run'
	from /Users/me/.rubies/ruby-2.2.4/lib/ruby/2.2.0/open3.rb:188:in `popen2e'
	from /Users/me/.rubies/ruby-2.2.4/lib/ruby/2.2.0/open3.rb:335:in `capture2e'
	from /Users/me/.gem/ruby/2.2.4/gems/rspec-core-3.4.1/lib/rspec/core/bisect/runner.rb:73:in `run_command'
	from /Users/me/.gem/ruby/2.2.4/gems/rspec-core-3.4.1/lib/rspec/core/bisect/runner.rb:65:in `block in run_locations'
	from /Users/me/.gem/ruby/2.2.4/gems/rspec-core-3.4.1/lib/rspec/core/bisect/server.rb:26:in `capture_run_results'
	from /Users/me/.gem/ruby/2.2.4/gems/rspec-core-3.4.1/lib/rspec/core/bisect/runner.rb:64:in `run_locations'
	from /Users/me/.gem/ruby/2.2.4/gems/rspec-core-3.4.1/lib/rspec/core/bisect/runner.rb:19:in `run'
	[trimmed]
```

RSpec only allows examples to be specified in the command-line options, instead of in option files (files/directories are [deleted](https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/configuration_options.rb#L157)), so using the command line arguments is the only way for `--bisect` to currently work.

I would like to put together a PR to address this, but wanted to discuss the appropriate solution first. Would it make sense to relax the restriction against files/directories in options files? Maybe have a bisect-specific options file that doesn't have the restriction?

**Comment by myronmarston** [comment]
Wow, you have quite the test suite!  And thanks for reporting this and offering to take a stab at fixing it. :heart: :heart: :heart:

I'm open to allowing files or directories to run in an option file, but I'm not sure it's the best solution.  For one, it's not something that generally makes sense (or that anyone has ever asked for).  Option files are for persistent options, not for one-off options for a specific run.  Allowing files or dirs in an option file would create complexity around how to deal with multiple lists of files or directories (e.g. from multiple option files or from an option file and from the command line).  Having the bisector hijack your options file also sounds kinda problematic.

Instead, I'd recommend we leverage the fact that the bisector uses a DRb server to communicate between the main process (the one you launched by running `rspec --bisect`) and the processes for the individual runs of a subset of the suite.  Currently there are two things that are communicated:

* [expected_failures](https://github.com/rspec/rspec-core/blob/996ddc1ca80cd6e77ff689395cb666c386dd3260/lib/rspec/core/bisect/server.rb#L46-L47) is _fetched_ by the sub process so that it can abort the run early once one of the expected failures passes or once all the expected failures fail.
* [latest_run_results](https://github.com/rspec/rspec-core/blob/996ddc1ca80cd6e77ff689395cb666c386dd3260/lib/rspec/core/bisect/server.rb#L49-L50) gets _set_ by the sub process -- it's the way it communicates the results of the run.

We could add `files_or_directories_to_run` to the list of things communicated over DRb.  Specifically, when a run starts with `--format bisect` (which indicates that is running a subset of the suite for a bisection run), have it get `files_or_directories_to_run` from the DRb server (similar to how it [gets expected_failures](https://github.com/rspec/rspec-core/blob/v3.4.0/lib/rspec/core/formatters/bisect_formatter.rb#L28)).  You'd just have to update [`Server#capture_run_results`](https://github.com/rspec/rspec-core/blob/v3.4.0/lib/rspec/core/bisect/server.rb#L23-L28) to take both `expected_failures` and `files_or_directories_to_run`.

Does that make sense?

**Comment by wendorf** [comment]
Yeah, absolutely. That sounds like a good approach. Thank you for the detailed response!

I'm traveling this week, so likely won't have much time to work on this, but when I return next week, I'll work on putting together a PR.

**Comment by skorfmann** [comment]
Just had the same error. Did you find some workaround (except for doing it manually)?

**Comment by myronmarston** [comment]
One work around is to reduce the number of specs.  You can run `rspec some/dir some/other/dir --bisect` to only bisect specs in particular directories.  Or you could temporarily delete some files (or some specs from some files) to remove them from consideration.

Of course, that requires that you know some specs that are not involved but it's worth a shot.

**Comment by nzgrover** [comment]
Having this issue as well, is there a fix on the horizon?

**Comment by myronmarston** [comment]
> is there a fix on the horizon?

As far as I know, no.  Someone will have to work up a fix.  It hasn't been a priority for anyone on the core team yet.  Are you interested in taking a stab, @nzgrover?

**Comment by Nitrodist** [comment]
I have a test suite that has about specs in it and I recently ran into this issue after upgrading from `3.3.1` to `3.4.3`. The test has some failing order dependent tests that we've been fixing over the last little while and after the upgrade the `--bisect` command doesn't work.

One quick hack to get around this issue and to also speed up fixing any order dependent tests is to run with `--fail-fast` -- this will stop the test suite after the first failure. So, say out of the 4000 tests that the first failure is after 200 tests have run, then the subsequent bisect commands will only reference the 199 passing tests plus the 1 failure at the end which lets you run the bisect command and is also *a lot* faster than running the rest of the suite.

**Comment by al2o3cr** [comment]
@myronmarston I have hacked together a fix locally (really really needed to bisect against a 9000+ example spec suite). Your comment from 12/31 was quite helpful, but the sticking point was that the piece of the "bisect client" Rspec (the one reading the specs-to-run from DRb) was a formatter which wasn't constructed until well after the specs-to-run were decided.

I got around that locally with shortcuts that would not make for reliable software (magic arguments & hard-coding). I've started digging into what it would take to make it work more cleanly.
